File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/tracker/ScreenState.java
package com.snowplowanalytics.snowplow.internal.tracker;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;

import com.snowplowanalytics.snowplow.internal.constants.Parameters;
import com.snowplowanalytics.snowplow.internal.constants.TrackerConstants;
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson;
import com.snowplowanalytics.snowplow.payload.TrackerPayload;
import com.snowplowanalytics.snowplow.internal.utils.Util;

@RestrictTo(RestrictTo.Scope.LIBRARY)
public class ScreenState implements State {
    private String name;
    private String type;
    private String id;
    private String previousName;
    private String previousId;
    private String previousType;
    private String transitionType;
    private String fragmentClassName;
    private String fragmentTag;
    private String activityClassName;
    private String activityTag;

    public ScreenState() {
        id = Util.getUUIDString();
        name = "Unknown";
    }

    @Nullable
    public String getPreviousId() {
        return previousId;
    }

    @Nullable
    public String getPreviousName() {
        return previousName;
    }

    @Nullable
    public String getPreviousType() {
        return previousType;
    }

    public synchronized void updateScreenState(@NonNull String id, @NonNull String name, @Nullable String type, @Nullable String transitionType) {
        this.populatePreviousFields();
        this.name = name;
        this.type = type;
        this.transitionType = transitionType;
        if (id != null) {
            this.id = id;
        } else {
            this.id = Util.getUUIDString();
        }
    }

    public synchronized void updateScreenState(@NonNull String id, @NonNull String name, @Nullable String type, @Nullable String transitionType, @Nullable String fragmentClassName, @Nullable String fragmentTag, @Nullable String activityClassName, @Nullable String activityTag) {
        this.updateScreenState(id, name, type, transitionType);
        this.fragmentClassName = fragmentClassName;
        this.fragmentTag = fragmentTag;
        this.activityClassName = activityClassName;
        this.activityTag = activityTag;
    }

    public void populatePreviousFields() {
        this.previousName = this.name;
        this.previousType = this.type;
        this.previousId = this.id;
    }

    @NonNull
    public SelfDescribingJson getCurrentScreen(boolean debug) {
        // this creates a screen context from screen state
        TrackerPayload contextPayload = new TrackerPayload();
        contextPayload.add(Parameters.SCREEN_ID, this.id);
        contextPayload.add(Parameters.SCREEN_NAME, this.name);
        contextPayload.add(Parameters.SCREEN_TYPE, this.type);
        if (debug) {
            contextPayload.add(Parameters.SCREEN_FRAGMENT, getValidName(fragmentClassName, fragmentTag));
            contextPayload.add(Parameters.SCREEN_ACTIVITY, getValidName(activityClassName, activityTag));
        }
        return new SelfDescribingJson(
                TrackerConstants.SCHEMA_SCREEN,
                contextPayload);
    }

    // Private methods

    private String getValidName(String s1, String s2) {
        if (s1 != null && s1.length() > 0) {
            return s1;
        }
        if (s2 != null && s2.length() > 0) {
            return s2;
        }
        return null;
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/core/tracker/ScreenState.kt
package com.snowplowanalytics.core.tracker

import androidx.annotation.RestrictTo
import com.snowplowanalytics.core.constants.Parameters
import com.snowplowanalytics.core.constants.TrackerConstants
import com.snowplowanalytics.core.utils.Util.uUIDString
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson
import com.snowplowanalytics.snowplow.payload.TrackerPayload

@RestrictTo(RestrictTo.Scope.LIBRARY)
class ScreenState : State {
    private var name: String
    private var type: String? = null
    private var id: String
    var previousName: String? = null
        private set
    var previousId: String? = null
        private set
    var previousType: String? = null
        private set
    private var transitionType: String? = null
    private var fragmentClassName: String? = null
    private var fragmentTag: String? = null
    private var activityClassName: String? = null
    private var activityTag: String? = null

    init {
        id = uUIDString()
        name = "Unknown"
    }

    @Synchronized
    fun updateScreenState(id: String?, name: String, type: String?, transitionType: String?) {
        populatePreviousFields()
        this.name = name
        this.type = type
        this.transitionType = transitionType
        if (id != null) {
            this.id = id
        } else {
            this.id = uUIDString()
        }
    }

    @Synchronized
    fun updateScreenState(
        id: String,
        name: String,
        type: String?,
        transitionType: String?,
        fragmentClassName: String?,
        fragmentTag: String?,
        activityClassName: String?,
        activityTag: String?
    ) {
        this.updateScreenState(id, name, type, transitionType)
        this.fragmentClassName = fragmentClassName
        this.fragmentTag = fragmentTag
        this.activityClassName = activityClassName
        this.activityTag = activityTag
    }

    private fun populatePreviousFields() {
        previousName = name
        previousType = type
        previousId = id
    }

    fun getCurrentScreen(debug: Boolean): SelfDescribingJson {
        // this creates a screen context from screen state
        val contextPayload = TrackerPayload()
        contextPayload.add(Parameters.SCREEN_ID, id)
        contextPayload.add(Parameters.SCREEN_NAME, name)
        contextPayload.add(Parameters.SCREEN_TYPE, type)
        if (debug) {
            contextPayload.add(
                Parameters.SCREEN_FRAGMENT,
                getValidName(fragmentClassName, fragmentTag)
            )
            contextPayload.add(
                Parameters.SCREEN_ACTIVITY,
                getValidName(activityClassName, activityTag)
            )
        }
        return SelfDescribingJson(
            TrackerConstants.SCHEMA_SCREEN,
            contextPayload
        )
    }

    // Private methods
    private fun getValidName(s1: String?, s2: String?): String? {
        if (s1 != null && s1.isNotEmpty()) {
            return s1
        }
        return if (s2 != null && s2.isNotEmpty()) {
            s2
        } else null
    }
}
